### ISSUE TITLE
Add push notifications support

### ISSUE THREAD

**Issue by dvor** [Antidote-for-Tox/Antidote#191]
toxcore should support relaying messages to push server (which can be also store offline messages?).

It is possible to send encrypted push notifications and decrypt them on device using PushKit.

See https://developer.apple.com/library/ios/documentation/Performance/Conceptual/EnergyGuide-iOS/OptimizeVoIP.html#//apple_ref/doc/uid/TP40015243-CH30-SW1

**Comment by Chuongv** [comment]
With Pushkit, we need a dedicated server that will run toxcore iterate. So are expecting users to host their own server with toxcore? It's a complicated problem considering the fact that Tox is a distributed system.

**Comment by GrayHatter** [comment]
@Chuongv is correct here. There's no reason you CANT do this, but you'd need to run a Tox instance for each client on the server.

**Comment by dvor** [comment]
Yeah, I know. After reading offline messages threads in toxcore issues I'm thinking of the following:
- user can choose some tox node (or several nodes) for storing offline messages
- mobile users can also subscribe to push notifications
- user will share those nodes with friends
- when user is offline friend will send encrypted messages to given nodes
- nodes will send push notifications if user was subscribed
- when user is back he'll ping node for messages and download them all.

I'll try to tweak toxcore and create proof of concept in nearest future, with more detailed explanation.

**Comment by dvor** [comment]
Also there is no need for running tox instance for each client. Single instance should be capable doing all that stuff.

**Comment by GrayHatter** [comment]
How will you separate messages and keys for the push server?

**Comment by dvor** [comment]
What exactly do you mean, can you explain a little bit more?

**Comment by GrayHatter** [comment]
How are you going to serve as a relay node for different users with one tox instance?

**Comment by dvor** [comment]
User will share his UserID (some identifier, TBD) with friend. Friend will send message + UserID to "offline" node. When user is back online he would be able to download all his messages using his UserID. That way single tox instance can maintain multiple clients.

Initially I was thinking of creating some proof of concept + proposal, but now I think I will do only proposal to discuss first. I'll structure everything and create a toxcore issue.

**Comment by GrayHatter** [comment]
@Dvor  you might want to take a look at https://docs.google.com/document/d/1op6zGR0KYdF7tTWSSX79KQieJu30vLZ6XG327kIBhxQ/edit

Also, that method seems messy. What's to stop any node from being malicious with the offline message?

**Comment by dvor** [comment]
Thanks, I'll take a look.

Malicious in what sense? The only bad thing node can do is to not deliver messages to user. In that case user can switch to another node.

Other malicious actions I see:

- node cannot read messages as they are encrypted
- node can send messages to another user, but that makes no sense as user won't be able to decrypt them
- node can send fake messages, that makes no sense as messages should be signed by friend.

**Comment by GrayHatter** [comment]
@dvor are clients expected to distribute their friends list to each push/relay node?

**Comment by dvor** [comment]
Nope. Friends wouldn't be connected to the node. They will send "friend request" like messages for offline storage.

I'm not sure if this will work correctly so I need proof of concept. Any remarks on that?

**Comment by GrayHatter** [comment]
@dvor the idea isn't bad, but how will you stop someone from overloading the message node

how will they know what users they can accept to/from?

how will my client decide which node to use?

how will my client know which node to get messages from?

whats' to stop a DDoS on nodes?

**Comment by dvor** [comment]
That's a good questions, I'm thinking about it. Some options how I see it:

> how will they know what users they can accept to/from?

User can share some secret key between node and his friend. This key will be used by node to accept messages and then give them back to user.

> how will my client decide which node to use?

From mobile perspective in any case there would be centralized part - Apple or Google push notifications service. In case if mobile user would like to receive notifications he will connect to any node that has ability to talk to Apple/Google. I'm thinking of having list of predefined nodes (like bootstrap ones) from which user can choose. In case if node is not performing well, he would be able to change it.

In non-notification case user can choose any node that supports offline storage. As a first approximation it will be done manually, I didn't thought about automating this process yet.

> how will my client know which node to get messages from?

Well, client will know nodes he subscribed to and will ask them for messages.

> whats' to stop a DDoS on nodes?

If node receives message without secret key from any user, it should drop it. For now I don't see any better way.

**Comment by GrayHatter** [comment]
so you'll your message nodes will accept a message from ANY user for EVERY client that uses them? How do you stop spam?

**Comment by dvor** [comment]
Spam on node or spam on user? Node will ignore messages that aren't related to any subscribed user. In case if message is for known user but there is no known secret key, message would be ignored as well and won't reach user.

Indeed, it is possible to DDoS node with messages which it will ignore immediately after reception.

**Comment by dvor** [comment]
Also there may be some storage limit for every friend/user after which all new messages would be ignored.

**Comment by dvor** [comment]
How bad is the idea? :-)

Actually I'm open to any solution which would bring push notifications to Tox. Without them iOS client is completely unusable.
Also maybe offline storage isn't necessary in that case - mobile device can get all messages from push notifications and store them. Somehow I missed that before.

**Comment by GrayHatter** [comment]
@dvor the problem is doing everything without trust, because in the end, the node WILL get compromised. So stored messages are now public, user / friend lists are now public, message meta data, everything the node sees.

From my point of view, if you're GOING to do push messaging, you'd need a node that will just forward the messages to the iOS client via PushKit. That's the least bad way.

But then you're using predetermined OTP keys to do the encryption, another crypto attack vector.

**Comment by dvor** [comment]
Okay, one more try. I tried to keep everything simple, offline storage is not a part of push notifications. @GrayHatter can you please check it and give some feedback?

-----

Alice - mobile user that wants to receive push notifications while being offline.
Bob - friend that wants to communicate with Alice while she is offline.
Charlie - another friend of Alice.
Server - server Alice decided to use as a push notifications relay. All communication with server are done using https.

## Registration

- Alice registers on Server to receive push notifications and receives back `<alice-id>` - an unique user id.

- For each friend Alice generates unique token and sends it to Server:

```
Bob - <bob-token>
Charlie - <charlie-token>
```

- When friend comes online, Alice sends him `<alice-id>`, friends generated token and Server url:

```
<alice-id>
<bob-token>
https://Server/whatever
```

## Normal flow

Both Bob and Alice are online.
Bob wants to send a message to Alice, he does that in normal way through toxcore.

## Push notifications

Bob is online, Alice is offline. Bob wants to send a message to Alice.

- Bob sends encrypted message, `<alice-id>`, `<bob-token>` to Server
- Server
  - checks if `<alice-id>` exists
  - checks if `<alice-id>` has `<bob-token>`
  - if everything is fine, sends push notification to Alice with encrypted message
  - responds to Bob with status 200.
- Alice receives a push notification and decrypts it.

## Removing friends

- Alice removes Charlie from friend list
- Alice also sends request to Server to remove `<charlie-token>`
- Charlie sends message, `<alice-id>`, `<charlie-token>` to Server
- Server
  - checks if `<alice-id>` exists
  - checks if `<alice-id>` has `<charlie-token>`
  - no `<charlie-token>` is found, nothing is being done
  - responds to Charlie with status 200.

## Changing Server

- Alice changes Server to Server2, receives new `<alice2-id>`
- When friend comes online, Alice sends him `<alice2-id>`, friends generated token and Server2 url. Friend replaces old values with new one.

## Few remarks

- Server doesn't know ToxID of either Alice, Bob and Charlie.
- There is not way for Charlie to know if push notification was actually send or not.
- For iOS device there is NO way to change deviceToken that is used for push notifications. This means that there is no way to change Server, as the old one would be able to send whatever it like. Because of that I'm planning to run my own server that will be hardcoded in Antidote client. This may be different on Android side.

**Comment by GrayHatter** [comment]
This isn't bad. 

Doesn't fire base push messaging work on iOS? With that you can change the device_token.

Does iPush have anything like tickle? Where you can just notify the device to reconnect? 

Can you come up with anything that doesn't' expose your whole friends list to the server?

**Comment by dvor** [comment]
@GrayHatter 

> Doesn't fire base push messaging work on iOS? With that you can change the device_token.

App gets device_token from iOS and there is no way to change it (well, it can change on device reset, after 2 years, etc.).

When using firebase it will be a Server I've described. iPhone will share device_token with firebase and there will be no way to unshare it /unsubscribe.

> Does iPush have anything like tickle? Where you can just notify the device to reconnect? 

This can be done, however it is very limited. I don't think that it will be usable for Tox client. From documentation:
> Your app has up to 30 seconds of wall-clock time to process the notification and call the specified completion handler block. In practice, you should call the handler block as soon as you are done processing the notification. The system tracks the elapsed time, power usage, and data costs for your app’s background downloads. Apps that use significant amounts of power when processing remote notifications may not always be woken up early to process future notifications.

-----

> Can you come up with anything that doesn't' expose your whole friends list to the server?

Basically friend list isn't exposed to server. All server knows is that some `<device_token>` has registered some friends with `<token1>`, `<token2>`, `<token3>`. ToxIDs of user or friends aren't exposed to the server.

Also Alice can register friends tokens whenever she like. She can register 100 tokens at once and then spend them when needed.

Another possible improvement is to give friend new token on each new session and discard the old ones:

- Alice registers 100 tokens: `<token1>`, `<token2>`, ...
- Bob goes online
- Alice shares `<token1>` with Bob
- Alice goes offline
- Alice goes online
- Alice shares `<token2>` with Bob, Alice discards `<token1>` on server.

**Comment by GrayHatter** [comment]
> Apps that use significant amounts of power when processing remote notifications may not always be woken up early to process future notifications.

Well then you're fucked... There's no way to wakeup tox without using a lot of data.

**Comment by GrayHatter** [comment]
How does the client advertise itself to the server? Just, here's a token, send this message, say it's from me?

How do you verify friends identities (message senders)? 

Is there any call back to notice the sender message was received? 

You said no, but I ask again because what if the message gets lost?

Can you explain more about how the messages gets from Sender to Receiver? I.e. what parts are tox, what parts are high availability servers, and what parts are Apple Inc servers?

**Comment by janniklorenz** [comment]
What about using this idea just as method to perform an abstract action after receiving an offline message? So the distributor of the app in the store just runs the server, which triggers the actual push notification, and PushKit triggers the distributor server.

So you can change the PushKit server easily.

**Comment by dvor** [comment]
@GrayHatter 

> What tokens would you use?

Tokens are just some unique strings. UUID can be used as a token, for example. I don't see any need for tokens to be generated by toxcore, it can be done by client.

Here is a detailed example of the flow:

- client gets device token from iOS
- client registers it's device token on server, gets client token: 
    `POST https://server/register | device_token` -> `client_token`

- client generates some friend tokens and registers them:
    `POST https://server/add_tokens | UUID1 UUID2 UUID3`

- friend comes online, client sends token to friend via toxcore:
    `tox_update_push_token(friend, "https://server/push", client_token, UUID1);`

- friend sends push notification:
    `POST https://server/push | <encrypted message> client_token UUID1`

- server receives notification, finds `device_token` that matches `client_token + UUID` and sends request to the Apple server.

- client removes outdated token
    `DELETE https://server/delete_token UUID1`


> How does the client advertise itself to the server? Just, here's a token, send this message, say it's from me?

Friend doesn't advertise itself. Just, here's client_token, here is my UUID1, send this message.

> How do you verify friends identities (message senders)? 

Friend will sign and encrypt message, client will be able to decrypt it and verify signature.

> Is there any call back to notice the sender message was received? 

Sender will only know that "everything is fine". This means that the message was processed by server. There is no way for sender to know if actual push notification was send - because receiver could already removed sender from contacts.

> You said no, but I ask again because what if the message gets lost?

Well, Apple doesn't guarantee that push notification will be delivered. I see push notifications as "notifying" service, not transport. This is a good question, I'll think more about it. My first guess is that sender should still send message via toxcore when receiver is online. So

- Alice goes offline
- Bob sends push notifications
- Alice sees notification, launches app
- when Bob and Alice are connected, Bob sends message in normal way

There will be an issue if Bob will already be gone. Offline messaging will solve that part (if it will be there one day :-) ). As I said I'll think more about this.

**Comment by dvor** [comment]
@janniklorenz if I understood you correctly, I as distributor will run server to which other servers will connect? Like this?

[friend] -> [some server that can be changed] -> [my server] -> [Apple] -> [device]

This can be done, but I don't see what is the benefit of increasing amount of intermediaries.

**Comment by janniklorenz** [comment]
Yes, exactly.

Since we cannot change the device_token, the only one who needs access to it is the distributor.
The App generates a token from the distributor server, which allows the PushKit server(actually in this case a bad name..) to use the distributor push service. So we can easily revoke it, when we change the PushKit server.

Also I wouldn't need an Apple Developer account to set up my own PushKit server.

**Comment by dvor** [comment]
@janniklorenz what is the benefit for you to run your own server?

The only thing I can think of is to hide your contacts from distributor. All he will know would be "send this message to this device". Is that's your goal?

**Comment by janniklorenz** [comment]
I think it's a good idea not to put everything into one piece.

In bigger scales of users, it may also be worth it to use multiple PushKits. Also different platforms can use the same thing with a different distributor server.

**Comment by dvor** [comment]
Some WIP https://github.com/anoadragon453/tox-offline-messaging-prop/pull/2
